ArduCopter/events.cpp

```diff
@@ -281,7 +281,14 @@ void Copter::failsafe_terrain_set_status(bool data_ok)
 void Copter::failsafe_terrain_on_event()
 {
     failsafe.terrain = true;
-    gcs().send_text(MAV_SEVERITY_CRITICAL,"Failsafe: Terrain data missing");
+    if (wp_nav->get_terrain_source() == AC_WPNav::TerrainSource::TERRAIN_FROM_TERRAINDATABASE) {
+        GCS_SEND_TEXT(MAV_SEVERITY_CRITICAL,"%s: %s data missing", "Failsafe", "Terrain");
+    } else if (wp_nav->get_terrain_source() == AC_WPNav::TerrainSource::TERRAIN_FROM_RANGEFINDER) {
+        GCS_SEND_TEXT(MAV_SEVERITY_CRITICAL,"%s: %s RNGFND Unhealthy", "Failsafe", "Terrain");
+    } else {
+        GCS_SEND_TEXT(MAV_SEVERITY_CRITICAL,"%s: %s", "Failsafe", "Terrain");
+    }
+
```

Comment:
```suggestion
    if (wp_nav->get_terrain_source() == AC_WPNav::TerrainSource::TERRAIN_FROM_TERRAINDATABASE) {
        GCS_SEND_TEXT(MAV_SEVERITY_CRITICAL, "Failsafe: Terrain%s", " data missing");
    } else if (wp_nav->get_terrain_source() == AC_WPNav::TerrainSource::TERRAIN_FROM_RANGEFINDER) {
        GCS_SEND_TEXT(MAV_SEVERITY_CRITICAL, "Failsafe: Terrain%s", " RNGFND Unhealthy");
    } else {
        GCS_SEND_TEXT(MAV_SEVERITY_CRITICAL, "Failsafe: Terrain%s", "");
    }
```
Is more the idea.

Reply:
Do you know if the linker will optimize strings across object files? If so "Failsafe" and "Terrain" seperately are more likely to save space.